src/base/utils/version.h

```diff
@@ -84,56 +89,75 @@ namespace Utils
         {
         }
 
+        constexpr bool isValid() const
+        {
+            return m_components.has_value();
+        }
+
         constexpr ComponentType majorNumber() const
         {
-            static_assert(N >= 1, "The number of version components is too small");
-            return m_components[0];
+            static_assert((N >= 1), "The number of version components is too small");
+
+            if (!isValid())
+                throw RuntimeError(u"Object is invalid"_qs);
+            return (*m_components)[0];
         }
 
         constexpr ComponentType minorNumber() const
         {
-            static_assert(N >= 2, "The number of version components is too small");
-            return m_components[1];
+            static_assert((N >= 2), "The number of version components is too small");
+
+            if (!isValid())
+                throw RuntimeError(u"Object is invalid"_qs);
+            return (*m_components)[1];
         }
 
         constexpr ComponentType revisionNumber() const
         {
-            static_assert(N >= 3, "The number of version components is too small");
-            return m_components[2];
+            static_assert((N >= 3), "The number of version components is too small");
+
+            if (!isValid())
+                throw RuntimeError(u"Object is invalid"_qs);
+            return (*m_components)[2];
         }
 
         constexpr ComponentType patchNumber() const
         {
-            static_assert(N >= 4, "The number of version components is too small");
-            return m_components[3];
+            static_assert((N >= 4), "The number of version components is too small");
+
+            if (!isValid())
+                throw RuntimeError(u"Object is invalid"_qs);
+            return (*m_components)[3];
         }
 
         constexpr ComponentType operator[](const std::size_t i) const
         {
-            return m_components.at(i);
+            if (!isValid())
+                throw RuntimeError(u"Object is invalid"_qs);
+            return m_components->at(i);
         }
 
         QString toString() const override
         {
+            if (!isValid())
+                return {};
```

Comment:
Why not throw?

Reply:
`QString {}` is also an invalid state which matches this conditional. Satisfying function signature should have priority over throwing exceptions.

Reply:
> `QString {}` is also an invalid state

It is just an "empty" string but it is still valid.

As for "invalid" Version instance. I think of it like the following. Since it takes place only as a workaround for compatibility with `QHash`, and it does not make sense for the implementation of the main behavior of the `Version` class, it would be logical to limit the possibility of using it only to the minimum of functions that are necessary for the purpose of its existence (i.e., ensuring compatibility with `QHash`). Otherwise, it may hide some logical errors in the code.

Reply:
>it would be logical to limit the possibility of using it only to the minimum of functions that are necessary for the purpose of its existence (i.e., ensuring compatibility with QHash). 

Do you mean remove `Version::isValid()` and all `isValid()` checking? and just ignore the invalid case?

Reply:
> > it would be logical to limit the possibility of using it only to the minimum of functions that are necessary for the purpose of its existence (i.e., ensuring compatibility with QHash).
> 
> Do you mean remove `Version::isValid()` and all `isValid()` checking? and just ignore the invalid state?

But then everyone will be free to use an invalid value in the code! I meant the opposite. Any use of an invalid value, except for those functions that are necessary for compatibility with QHash, should cause an error so that we can more easily detect incorrect code.

Reply:
Well , maybe I'm just suffering from perfectionism right now. But I'm really confused by the fact that it looks like mixing different ways/patterns. If we do not allow the creation of invalid instances in some cases, then why do we allow in others? Well, we found out that we need a default constructor for compatibility with QHash, but it cannot create valid instances in the general case, since it violates the requirement for Mandatory number of elements. Well, but since such invalid values are not intended for direct use in the code, then how to avoid such use?


On the other hand, some aspects of this class look too far-fetched. Take Mandatory. I don't see any benefit from it at compile time. Moreover, I would limit its influence only on the conversion of the version to a string, because even for parsing it looks superfluous. I don't think that in the real world we can expect that the version of a particular product will always be written respecting it.
So why not just simplify things by removing redundant restrictions?

Reply:
Well I already have a version in the work.
I will change `isValid()` to be only checking for meaningful version numbers, not representing some state of the Version object. For a Version to be meaningful, all numbers should be non-negative and at least one number will be greater than 0. Therefore {-1,-1,-1} is constructable but it is not valid/meaningful. Other observer/getter functions will not be influenced by `isValid()`.

As for Mandatory, I'm not prepared to deal with it yet so I'll defer removing it.

Reply:
> As for Mandatory, I'm not prepared to deal with it yet so I'll defer removing it.

Ok.

Reply:
I pushed another commit.